Allow extra descriptor sources in reflection mode

alexanderchiu opened the pull request:
Allows descriptors or proto files to be supplied that are used as an
additional descriptor source when server reflection is used. This is
needed to resolve message types, embedded in an Any message, that the
server has no knowledge of.

Resolves: #135

alexanderchiu commented:
There is also a test server here: https://github.com/alexanderchiu/grpcurl/tree/alex/test-server which I haven't included in the PR.

alexanderchiu commented:
- [ ] desc_source.go:272:3: should replace loop with exts = append(exts, addExt...)

alexanderchiu commented:
Sounds good to me. A `-use-reflection` flag will be a lot cleaner than the current mechanism. Will make the proposed changes soonest.

jhump commented on cmd/grpcurl/grpcurl.go:
```diff
@@ -225,6 +231,11 @@ func main() {
 		warn("The -emit-defaults is only used when using json format.")
 	}
 
+	// No protoset or protofiles provided, Use reflection
```
At this point, we've already parsed all flags. If there was a `-reflection=false`, which is a condition worth failing with an error message if there were no protos or protosets, we wouldn't know it and would just be overwriting the option.

If you're not sure how to address this (it would require creating a custom implementation of `flag.Value` to track both whether a value was set on the command-line and what that value was), don't worry about it. You can just add a TODO, and I can follow-up with some small cleanup and input validation later.

jhump commented on desc_source.go:
```diff
@@ -206,8 +206,33 @@ func DescriptorSourceFromServer(_ context.Context, refClient *grpcreflect.Client
+// DescriptorSourceFromServerWithAdditionalProtosets creates a DescriptorSource
+// that uses the given gRPC reflection client to interrogate a server for descriptor
+// information and the supplied protosets as a source of additional symbols.
+func DescriptorSourceFromServerWithAdditionalProtosets(_ context.Context, refClient *grpcreflect.Client, fileNames ...string) DescriptorSource {
```
Sorry I didn't leave this comment in earlier iteration, before you re-worked the flags. But I don't think we want to widen the API like this. Instead, I think you could have a custom implementation of `DescriptorSource` in the `main` package that just had two underlying sources -- reflection source, then the proto or protoset source.

Its implementation would just look at the reflection source for listing services, but could then use both when resolving symbols (consult first then fallback to other on error).

alexanderchiu commented:
> I think we could implement this without adding the extra types and methods to the `grpcurl` package.
> 
> If we find them to be valuable for others, I think we'd add them to the exported package as generic composition functions, instead of bespoke "reflection plus protos" sources.

Thanks for all the feedback. I've learned quite a bit through the process. I've updated the PR with

1. Reverted the changes in the grpcurl package and added an implementation of `DescriptorSource` in the main package.
2. Fixed the `-use-reflection` flag handling including throwing an error when `-use-reflection` is explicitly set to false and no protos/protosets are provided.

alexanderchiu commented on cmd/grpcurl/grpcurl.go:
```diff
@@ -672,3 +722,13 @@ func writeProtoset(descSource grpcurl.DescriptorSource, symbols ...string) error
 	return grpcurl.WriteProtoset(f, descSource, symbols...)
 }
+
+func flagSet(name string) bool {
```
Thanks for the suggestion. Made these changes.

alexanderchiu commented on cmd/grpcurl/grpcurl.go:
```diff
@@ -181,6 +187,33 @@ func (s *multiString) Set(value string) error {
+
+func (cs compositeSource) AllExtensionsForType(typeName string) ([]*desc.FieldDescriptor, error) {
+	exts, err := cs.reflection.AllExtensionsForType(typeName)
+	if err == nil {
```
I think this should be handled now.
